src/test/java/com/zendesk/maxwell/producer/InflightMessageListTest.java

```diff
@@ -54,4 +59,36 @@ public void testOutOfOrderComplete() {
 		ret = list.completeMessage(p1);
 		assertEquals(p3, ret);
 	}
+
+	@Test
+	public void testAddMessageWillWaitWhenCapacityIsFull() throws InterruptedException {
+		AddMessage addMessage = new AddMessage();
+		new Thread(addMessage).start();
+		assertThat("Should never exceed capacity", list.size(), lessThanOrEqualTo(capacity));
+
+		long wait = 500;
+		Thread.sleep(wait);
+		list.completeMessage(p1);
+
+		Thread.sleep(100);
+		assertThat("Should never exceed capacity", list.size(), lessThanOrEqualTo(capacity));
+		long elapse = addMessage.end - addMessage.start;
+		assertThat("Should have waited message to be completed", elapse, greaterThan(wait));
+	}
+
+	class AddMessage implements Runnable {
+		long start;
+		long end;
+
+		@Override
+		public void run() {
+			start = System.currentTimeMillis();
+			try {
+				list.addMessage(p4);
+			} catch (InterruptedException e) {
+				e.printStackTrace();
```

Comment:
Can you confirm that the test fails if this thread dies with (say) a RuntimeExecption? And if so, maybe just raise one here instead of logging, because this should never happen.

Reply:
If `InterruptedException` is thrown during `wait()`, the rest behaviour will be aborted which means it won't add the message. I've changed the assertion to check the list size should be exactly equal to the capacity to make sure the test will fail for this case. Any other exceptions will blow up the test anyways.